test/integration/api/FactoryAPI.spec.ts

```diff
@@ -388,4 +391,234 @@ describe('FactoryAPI', () => {
+        const expectedResult = new BigNumber(10).pow(18);
+
+        const result = await subject();
+
+        expect(result).to.bignumber.equal(expectedResult);
+      });
+    });
+
+    describe('when the decimals of the components are all different', async () => {
+      const smallerDecimal = 8;
+      const largerDecimal = 18 - smallerDecimal;
+      let decimalsList: number[];
+
+      beforeAll(async () => {
+        const tokenCount = 2;
+        decimalsList = [largerDecimal, smallerDecimal];
+        componentInstances = await deployTokensSpecifyingDecimals(tokenCount, decimalsList, provider);
+      });
+
+      afterAll(async () => {
+        componentInstances = [];
+      });
+
+      test('it calculates the min minimum natural unit correctly', async () => {
+        const expectedResult = new BigNumber(10).pow(largerDecimal);
+
+        const result = await subject();
+
+        expect(result).to.bignumber.equal(expectedResult);
+      });
+    });
+  });
+
+  describe('calculateRequiredComponentUnits', async () => {
+    let subjectComponentPrices: BigNumber[];
+    let subjectComponents: StandardTokenMockContract[];
+    let subjectComponentAllocations: BigNumber[];
+    let subjectTargetSetPrice: BigNumber;
+
+    beforeEach(async () => {
+      const tokenCount = 2;
+      const decimalsList = [18, 18];
+      subjectComponents = await deployTokensSpecifyingDecimals(tokenCount, decimalsList, provider);
+      subjectComponentPrices = [new BigNumber(2), new BigNumber(2)];
+      subjectComponentAllocations = [new BigNumber(0.5), new BigNumber(0.5)];
+      subjectTargetSetPrice = new BigNumber(10);
+    });
+
+    async function subject(): Promise<BigNumber[]> {
+      const subjectComponentAddresses = _.map(subjectComponents, components => components.address);
+
+      return await factoryAPI.calculateRequiredComponentUnits(
+        subjectComponentPrices,
+        subjectComponentAddresses,
+        subjectComponentAllocations,
+        subjectTargetSetPrice,
+      );
+    }
+
+    describe('when the inputs are standard', async () => {
+      test('should calculate the correct required component units', async () => {
+        const expectedResult = [
+          new BigNumber(2.5).mul(ether(1)),
+          new BigNumber(2.5).mul(ether(1)),
+        ];
+
+        const result = await subject();
+
+        expect(JSON.stringify(result)).to.equal(JSON.stringify(expectedResult));
+      });
+    });
+
+    describe('when the allocation inputs do not sum up to 1', async () => {
+      beforeEach(async () => {
+        subjectComponentAllocations = [new BigNumber(0.5), new BigNumber(0.49)];
+      });
+
+      test('it should throw', async () => {
+        return expect(subject()).to.be.rejectedWith(`The component percentages inputted do not add up to 1`);
+      });
+    });
+
+    describe('when the inputs represent a real world integration case', async () => {
```

Comment:
What's `real world integration case`? Can you specify in short which tokens/units each refers to?

Reply:
Can you explain what price you're trying to achieve?